lang/src/org/partiql/lang/eval/ExprValueExtensions.kt

```diff
@@ -377,13 +379,25 @@ private fun Number.toLongFailingOverflow(locationMeta: SourceLocationMeta?): Lon
     }
 }
 
-
-
 /**
  * An Unknown value is one of `MISSING` or `NULL`
  */
 internal fun ExprValue.isUnknown() : Boolean = this.type.isUnknown
 /**
  * The opposite of [isUnknown].
  */
-internal fun ExprValue.isNotUnknown() : Boolean = !this.type.isUnknown
\ No newline at end of file
+internal fun ExprValue.isNotUnknown() : Boolean = !this.type.isUnknown
+
+/**
+ * Creates a filter for unique ExprValues consistent with exprEquals. This filter is stateful keeping track of 
+ * seen [ExprValue]s.
+ * 
+ * This filter is **stateful**!
+ * 
+ * @return false if the value was seen before 
+ */
+internal fun uniqueExprValueFilter(): (ExprValue) -> Boolean {
+    val seen = TreeSet<ExprValue>(DEFAULT_COMPARATOR)
```

Comment:
hmm...not a HashSet?

Reply:
`ExprValue` uses the default java equals and hashcode impl not so we have to use a TreeSet to use the same equality definition as the `=` PartiQL operator which is the `DEFAULT_COMPARATOR`

Reply:
Don't agree 100% with David here. We have an `equals` definition for `ExprValue`, the comparator, so the question really is implementing a hashcode which is non trivial but far from impossible. 

The complex part is that ExprValue is an interface and both equals and hashcode must be built on top of the interface itself for consistency, apart from the fact that an ExprValue may be a sequence or a struct with lots of nesting. 

IMHO we may end up implementing a hashcode if there is a compeling reason such as performance for GroupBy and Distinct.